mlflow/pyfunc/__init__.py

```diff
@@ -78,6 +87,125 @@
+**pyfunc** flavor in a variety of machine learning frameworks (Scikit-learn, Keras, Pytorch, and
+more); however, they do not cover every use case. For example, you may want to create an MLflow
+model with the **pyfunc** flavor using a framework that MLflow does not natively support.
+Alternatively, you may want to build an MLflow model that executes custom logic when evaluating
+queries, such as preprocessing and postprocessing routines. Therefore, ``mlflow.pyfunc``
+provides utilities for creating **pyfunc** models from arbitrary code and model data.
+
+The :meth:`save_model()` and :meth:`log_model()` methods are designed to support multiple workflows
+for creating custom **pyfunc** models that incorporate custom inference logic and data dependencies.
+Data dependencies are **parameters** and **artifacts** that the inference logic may use to
+evaluate queries. These components are defined as follows:
+
+1. **Parameters**: Python objects. For example, a numpy matrix containing precomputed feature
+   values is a **parameter**.
+2. **Artifacts**: Files or directories, such as serialized Python models or CSVs. For example, a
+   serialized Tensorflow graph is an **artifact**; an MLflow model directory is also an
+   **artifact**.
+
+.. _pyfunc-create-custom-workflows:
+
+Workflows
+#########
+
+:meth:`save_model()` and :meth:`log_model()` support the following workflows:
+
+    1. Collecting **artifacts** and **parameters** to create an MLflow model
+
+       Given a set of **artifact** URIs and **parameters**,
+       :meth:`save_model()` and :meth:`log_model()` can automatically download **artifacts** from
+       their URIs and serialize **parameters** to create an MLflow model directory. **artifacts**
+       and **parameters** are specified by the ``artifacts`` and ``parameters`` method arguments,
+       respectively.
+
+       In this case, you must provide a Python class, called a **model class**, with two
+       methods:
+
+           - ``__init__()``: This method loads and interprets **artifacts** from the MLflow
+             model directory.
+
+             *Note:* **parameters** are *automatically* read from the MLflow model directory as
+             Python objects.
+
+           - ``predict()``: This method leverages **artifacts** and **parameters** to evaluate
+             queries. It must adhere to the :ref:`pyfunc-inference-api`.
+
+       The **model_class** must inherit from :class:`~PythonModel`. It is specified via the
+       ``model_class`` parameter and is automatically serialized and deserialized as a Python
+       class object.
+
+    2. Importing data as an MLflow model
+
+       If you already have a collection of model data, :meth:`save_model()` and :meth:`log_model()`
+       can import the data as an MLflow model. Model data may consist of both **artifacts** and
+       *serialized* **parameters**. The ``data_path`` argument specifies the local filesystem path
```

Comment:
I wonder if instead of `parameters`, another name like `data` or `model_data` would be more obvious. If I conceptualize something as a file (e.g., customers.csv), I would probably pass in the name "customers.csv" to parameters, as opposed to the result of `open("customers.csv").read()`.

Reply:
After offline discussion, I've removed `parameters` and replaced `model_class` with `python_model`, where the latter term now refers to a python class **instance**, rather than a class **type**. This instance is now serialized and deserialized.